src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -198,19 +198,43 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 let place = &self.move_data.move_paths[mpi].place;
 
                 let ty = place.ty(self.mir, self.infcx.tcx).to_ty(self.infcx.tcx);
-                let note_msg = match self.describe_place_with_options(
-                    place,
-                    IncludingDowncast(true),
-                ) {
-                    Some(name) => format!("`{}`", name),
+                let opt_name = self.describe_place_with_options(place, IncludingDowncast(true));
+                let note_msg = match opt_name {
+                    Some(ref name) => format!("`{}`", name),
                     None => "value".to_owned(),
                 };
-
-                err.note(&format!(
-                    "move occurs because {} has type `{}`, \
-                     which does not implement the `Copy` trait",
-                    note_msg, ty
-                ));
+                let mut note = true;
+                for decl in &self.mir.local_decls {
+                    if decl.ty == ty && decl.name.map(|x| x.to_string()) == opt_name {
```

Comment:
I think we can do this much more efficiently. The variable `place` contains the, well, *place* which was moved -- in this case, we are specifically interested in moves from a local variable, right? So we can just do something like:

```rust
if let Place::Local(local) = place {
  let decl = &self.mir.local_decls[local];
}
```

Besides being inefficient, this check is imprecise. If you for example have a shadow of a variable, there could be a bug I think (maybe worth adding a test):

```rust
fn main() {
    let x: Vec<u32> = vec![1, 2, 3];
    drop(x);

    let x: Vec<u32> = vec![1, 2, 3];
    drop(x);
    drop(x);
}
```

I think that will highlight the wrong variable.

Reply:
Fixed.